controlplane/src/core/controllers/scim.ts

```diff
@@ -0,0 +1,408 @@
+        const keycloakUsers = await opts.keycloakClient.client.users.find({
+          realm: opts.keycloakRealm,
+          email: member.email,
+        });
+        if (keycloakUsers.length === 0) {
+          continue;
+        }
+        users.push({
+          id: member.userID,
+          email: member.email,
+          userName: member.email,
+          active: keycloakUsers[0].enabled || true,
+        });
+      }
+    }
+
+    return res.code(200).send({
+      schemas: ['urn:ietf:params:scim:api:messages:2.0:ListResponse'],
+      totalResults: users.length,
+      startIndex,
+      itemsPerPage: users.length,
+      Resources: users,
+    });
+  });
+
+  fastify.get('/Users/:userID', async (req: FastifyRequest<{ Params: { userID: string } }>, res) => {
+    const authContext = req.authContext;
+
+    const userID = req.params.userID;
+
+    const user = await opts.organizationRepository.getOrganizationMember({
+      organizationID: authContext.organizationId,
+      userID,
+    });
+
+    if (!user) {
+      return res.code(404).send({
+        schemas: ['urn:ietf:params:scim:api:messages:2.0:Error'],
+        detail: 'User not found',
+        status: 404,
+      });
+    }
+
+    const keycloakUsers = await opts.keycloakClient.client.users.find({
+      realm: opts.keycloakRealm,
+      email: user.email,
+    });
+    if (keycloakUsers.length === 0) {
+      return res.code(404).send({
+        schemas: ['urn:ietf:params:scim:api:messages:2.0:Error'],
+        detail: 'User not found',
+        status: 404,
+      });
+    }
+
+    return res.code(200).send({
+      schemas: ['urn:ietf:params:scim:schemas:core:2.0:User'],
+      id: user.userID,
+      userName: user.email,
+      name: {
+        givenName: keycloakUsers[0].firstName || '',
+        middleName: '',
+        familyName: keycloakUsers[0].lastName || '',
+      },
+      active: keycloakUsers[0].enabled || true,
+      emails: [
+        {
+          primary: true,
+          value: user.email,
+          type: 'work',
+        },
+      ],
+      groups: [],
+      meta: {
+        resourceType: 'User',
+      },
+    });
+  });
+
+  // create a user
+  fastify.post(
+    '/Users',
+    async (
+      req: FastifyRequest<{
+        Body: {
+          schemas: string[];
+          userName: string;
+          name: { givenName: string; familyName: string };
+          emails: { primary: boolean; value: string; type: string }[];
+          password: string;
+          displayName: string;
+          groups: string[];
+          active: boolean;
+          locale: string;
+          externalId: string;
+        };
+      }>,
+      res,
+    ) => {
+      const authContext = req.authContext;
+
+      const { userName, name, emails, password, displayName, groups, active, locale, externalId } = req.body;
+
+      const email = emails.find((e) => e.primary === true)?.value || userName;
+
+      const user = await opts.userRepository.byEmail(email);
```

Comment:
We only store the email of the user that he used for signing up but happen when he uses a linked account? I think we should ask keycloak if a user already exists.

Reply:
I do check